### PR TITLE
refactor: incremental update of inlay hints and more options

### PR DESCRIPTION
**Pull request by yioneko**
Previously, the inlay hints are dropped and requested as a whole on every update of text. This will cause apparent lag in large file, and tsserver is even less responsive as it's busy dealing with the large amounts of inlay hints. By requesting and updating the hints incrementally, the performance will be much better.

Three extra options are exposed to help customization. 
- `inlay_hints_priority` is used to resolve the conflict with other extmarks set on the same line. 
- `inlay_hints_throttle` configures the throttle time of requests as the text may change much frequently.
- `inlay_hints_format` allow user to customize the highlight and text of different kind of hints, because the original text format of hint is confusing when messed up on the single line (`param: ` and `: return`).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -170,6 +170,18 @@ lspconfig.tsserver.setup({
             -- inlay hints
             auto_inlay_hints = true,
             inlay_hints_highlight = "Comment",
+            inlay_hints_priority = 200, -- priority of the hint extmarks
+            inlay_hints_throttle = vim.o.updatetime, -- throttle the inlay hint request, set to 0 for no throttle
```
Comment (jose-elias-alvarez) [inline README.md:174]: Do you think there's a use case for exposing this separately? I think using `updatetime` is totally sensible, and I can't imagine many users would want to change it.

File: lua/nvim-lsp-ts-utils/inlay-hints.lua [hunk 2]
```diff
@@ -2,99 +2,156 @@ local o = require("nvim-lsp-ts-utils.options")
...
-        "autocmd_fun()"
-    )
+local ns = api.nvim_create_namespace("ts-inlay-hints")
+M.ns = ns
+
+-- enabled[bufnr] = false
+M.enabled = {}
+
+local function buf_enabled(bufnr)
+    if bufnr ~= nil then
```
Comment (jose-elias-alvarez) [inline lua/nvim-lsp-ts-utils/inlay-hints.lua:15]: Since we repeat it in a few places, could we extract the `bufnr` resolution logic into a helper function, kind of like how Neovim does it [here](https://github.com/neovim/neovim/blob/9804a2870f6f308f788f939f52958e3fbd2adaac/runtime/lua/vim/lsp.lua#L75)?

File: lua/nvim-lsp-ts-utils/inlay-hints.lua [hunk 3]
```diff
@@ -2,99 +2,156 @@ local o = require("nvim-lsp-ts-utils.options")
...
+        if bufnr == 0 then
+            bufnr = api.nvim_get_current_buf()
+        end
+        if M.enabled[bufnr] == nil then
+            M.enabled[bufnr] = false
+        end
+        return M.enabled[bufnr]
```
Comment (jose-elias-alvarez) [inline lua/nvim-lsp-ts-utils/inlay-hints.lua:22]: If we strictly want to return a boolean (which I don't think is necessary) couldn't we just return `M.enabled[bufnr] ~= nil`?

File: lua/nvim-lsp-ts-utils/inlay-hints.lua [hunk 4]
```diff
@@ -2,99 +2,156 @@ local o = require("nvim-lsp-ts-utils.options")
...
                     value,
                 }
             end
         end
 
-        local lines = api.nvim_buf_get_lines(bufnr, 0, -1, false)
-        for key, value in pairs(parsed) do
-            local line = tonumber(key)
-            if lines[line + 1] then
+        local lines_cnt = vim.fn.line("$")
```
Comment (jose-elias-alvarez) [inline lua/nvim-lsp-ts-utils/inlay-hints.lua:72]:
```suggestion
        local lines_cnt = #api.nvim_buf_get_lines(bufnr, 0, -1, false)
```

Just a personal preference, but I prefer to avoid the Vim function, which is hard to understand if you're not familiar with Vimscript.

File: lua/nvim-lsp-ts-utils/inlay-hints.lua [hunk 5]
```diff
@@ -2,99 +2,156 @@ local o = require("nvim-lsp-ts-utils.options")
...
+            bufnr = api.nvim_get_current_buf()
+        end
+        if M.enabled[bufnr] == nil then
+            M.enabled[bufnr] = false
+        end
+        return M.enabled[bufnr]
+    end
+end
+
+local function set_enabled(bufnr, enabled)
```
Comment (jose-elias-alvarez) [inline lua/nvim-lsp-ts-utils/inlay-hints.lua:26]: I feel like this would be a bit cleaner if we separate it into `enable` and `disable` functions.

**Review by jose-elias-alvarez — COMMENTED** [review]
I left a few minor comments but this looks great! I definitely feel the improvement in responsiveness. I'll continue testing this out locally.

File: README.md (hunk 1 above)
Comment (yioneko) [reply README.md:174]: I originally expose this option for users who want to disable the throttled behavior like the previous implementation, if they encounter any problems after upgrade. If the throttling works fine, we can certainly remove it.

File: lua/nvim-lsp-ts-utils/inlay-hints.lua (hunk 4 above)
Comment (yioneko) [reply lua/nvim-lsp-ts-utils/inlay-hints.lua]: Changed to `api.nvim_buf_line_count`. I didn't know this function exists while developing 😅

File: README.md (hunk 1 above)
Comment (jose-elias-alvarez) [reply README.md:174]: In that case I think it's better to test these changes a bit more before merging. I'd prefer not to expose an option that we'll remove in the future, since it would qualify as a breaking change.

File: lua/nvim-lsp-ts-utils/inlay-hints.lua [hunk 6]
```diff
@@ -2,99 +2,162 @@ local o = require("nvim-lsp-ts-utils.options")
...
 end
 
-local function hide(bufnr)
-    api.nvim_buf_clear_namespace(bufnr, M.state.ns, 0, -1)
+local function set_buf_enabled(bufnr)
+    M.enabled[resolve_bufnr(bufnr)] = true
+end
+
+local function set_buf_disabled(bufnr)
+    M.enabled[resolve_bufnr(bufnr)] = false
```
Comment (jose-elias-alvarez) [inline lua/nvim-lsp-ts-utils/inlay-hints.lua:33]:
```suggestion
    M.enabled[resolve_bufnr(bufnr)] = nil
```

More of a housekeeping thing, but I think it's better to clear this completely for garbage collection.

File: README.md [hunk 7]
```diff
@@ -170,6 +170,18 @@ lspconfig.tsserver.setup({
...
+            inlay_hints_format = { -- format options for individual hint kind
+                Type = {
+                    highlight = nil,
+                    text = function(text)
```
Comment (jose-elias-alvarez) [inline README.md:178]: I like the idea of exposing this option, but I personally find the arrow syntax a little confusing (the distinction between `->` and `=>` is a little hard to parse at a glance). Is this based on something or just personal preference? If it's the latter, I think I'd prefer to leave this as `nil` by default but document the arrow as a potential option.

Comment (yioneko) [reply README.md:178]: It might depend on the font used. It looks easy to distinguish between them for the font I use. I originally think it's a better default but since that's not good enough we should simply document that as an example.

Comment (jose-elias-alvarez) [reply README.md]: This is what it looks like using PragmataPro:

Screen Shot 2021-12-29 at 1 41 01 PM

And this is with Iosevka:

Screen Shot 2021-12-29 at 1 40 46 PM

It definitely looks better with wider fonts, but these are the only 2 fonts I use regularly, so I think it's best to stick with the default.

File: README.md [hunk 8]
```diff
@@ -88,7 +88,14 @@ built-in LSP client.
   - `inlay_hints_highlight` (string): Highlight group used for inlay hints.
     Defaults to "Comment".
 
-- Filter `tsserver` diagnostics
+  - `inlay_hints_priority` (number): Priority of the hint extmarks. Change
```
Comment (jose-elias-alvarez) [inline README.md:91]: Could we also document that `updatetime` affects the time between updating inlay hints?

**Review by jose-elias-alvarez — COMMENTED** [review]
LGTM! I'll plan on testing this for the rest of the week and merge early next week if there's no issues.

File: README.md (hunk 1 above)
Comment (yioneko) [reply README.md]: After looking at `:h updatetime` I think I misunderstood the usage of this option. The option defaults to a quite large value `4000` and is used by nvim internally to determine the frequency of calling `fsync()`. So it may be better to set the throttle time to a rather low value (200, etc.) and expose the option to let users modify it if they want tsserver to respond either faster or slower.

Comment (yioneko) [reply README.md]: I think we can also lift the option to plugin-level, for example, named as `delay` in order to be used by other module if possible.

Comment (jose-elias-alvarez) [reply README.md]: I see, I had thought that other plugins were also using `updatetime`, but that's because it also affects `CursorHold` (which isn't related to this). I'm always hesitant to add more options, but it's fine in this case (and I think we should keep it local to this specific functionality, since we never know how `tsserver` will behave in different environments).

**Comment by jose-elias-alvarez** [comment]
I haven't done a ton of JavaScript work lately but this seems to be working well. Let me know what you think about the config issue, but other than that I think this is okay to merge.

**Comment by yioneko** [comment]
Sorry for late response, previously I'm busy with personal stuff but these days I might have enough time to pick it up.

I think we should not rely on `updatetime` and bring the throttle time option back since there is an issue for `CursorHold`: https://neovim.discourse.group/t/psa-dont-rely-on-cursorhold-yet/1813. Also think setting default to 150ms is more reasonable since nvim now use it as default debounce time for lsp servers: https://github.com/neovim/neovim/pull/16908

And the another thing worth notice is that the changes introduce in https://github.com/jose-elias-alvarez/nvim-lsp-ts-utils/commit/4655fbcb85b55bc34c80aa3e9400ca069b3f1363 makes the throttle behavior actually becomes debounced since the on_lines events are accumulated and only the last response will be accepted and handled. I'm not sure whether it will be alright or better be reverted.

**Comment by jose-elias-alvarez** [comment]
> Sorry for late response, previously I'm busy with personal stuff but these days I might have enough time to pick it up.
> 
> I think we should not rely on `updatetime` and bring the throttle time option back since there is an issue for `CursorHold`: https://neovim.discourse.group/t/psa-dont-rely-on-cursorhold-yet/1813. Also I think setting default to 150ms is more reasonable since nvim now use it as default debounce time for lsp servers: [neovim/neovim#16908](https://github.com/neovim/neovim/pull/16908)
> 
> And the another thing worth notice is that the changes introduce in [4655fbc](https://github.com/jose-elias-alvarez/nvim-lsp-ts-utils/commit/4655fbcb85b55bc34c80aa3e9400ca069b3f1363) makes the throttle behavior actually becomes debounced since the on_lines events are accumulated and only the last response will be accepted and handled. I'm not sure whether it will be alright or better be reverted.

Yep, definitely agreed about `updatetime` and the default setting. I don't think debouncing changes is a big deal (it might reduce flickering, actually) so I'm happy with the way things are working now.

**Comment by jose-elias-alvarez** [comment]
Everything looks good to me, and I've been using this without issues. Let me know when you think it's ready and I'll merge it.

**Comment by yioneko** [comment]
Sure, I think there is no other issue.

**Comment by jose-elias-alvarez** [comment]
Great – thank you for your work on this!
